### PR TITLE
[BIOMAGE-1759] Test cellsets patching through the API.

### PR DESCRIPTION
**Pull request by gerbeldo**
# Description
We are patching cellsets through the API, but currently don't have any tests for that functionality. The goal is to add them.

# Details
#### URL to issue
https://biomage.atlassian.net/browse/BIOMAGE-1759

#### Link to staging deployment URL (or set N/A)
N/A

#### Links to any PRs or resources related to this PR
https://github.com/hms-dbmi-cellenics/worker/pull/266

#### Integration test branch
master

# Merge checklist
Your changes will be ready for merging after all of the steps below have been completed.

### Code updates
Have best practices and ongoing refactors being observed in this PR
- [ ] Migrated any selector / reducer used to the new format.

### Manual/unit testing
- [x] Tested changes using InfraMock locally **or** no tests required for change, e.g. Kubernetes chart updates.
- [x] Validated that current unit tests for code work as expected and are sufficient for code coverage **or** no unit tests required for change, e.g. documentation update.
- [x] Unit tests written **or** no unit tests required for change, e.g. documentation update.

### Integration testing
**You must check the box below** to run integration tests on the latest commit on your PR branch.
Integration tests have to pass before the PR can be merged. Without checking the box, your PR
**will not pass** the required status checks for merging.

- [x] Started end-to-end tests on the latest commit.

### Documentation updates
- [x] Relevant Github READMEs updated **or** no GitHub README updates required.
- [x] Relevant Wiki pages created/updated **or** no Wiki updates required.

### Optional
- [ ] Staging environment is unstaged before merging.
- [X] Photo of a cute animal attached to this PR.

image

### CODE REVIEW

File: pipeline-runner/R/qc-7-embed_and_cluster.R [hunk 1]
```diff
@@ -3,63 +3,90 @@
...
-embed_and_cluster <- function(scdata, config, sample_id, cells_id, task_name = "configureEmbedding") {
-  message("starting clusters")
-  clustering_method <- config$clusteringSettings$method
-  methodSettings <- config$clusteringSettings$methodSettings[[clustering_method]]
-  message("Running clustering")
-  cellSets <- runClusters(clustering_method, methodSettings$resolution, scdata)
-  message("formatting cellsets")
-  formated_cell_sets <- format_cell_sets_object(cellSets, clustering_method, scdata@misc$color_pool)
-  message("udating through api")
+embed_and_cluster <-
```
Comment (ivababukova) [inline pipeline-runner/R/qc-7-embed_and_cluster.R:6]: it looks like nothing got changed in this file apart from formatting -- is that right?

Comment (gerbeldo) [reply pipeline-runner/R/qc-7-embed_and_cluster.R:6]: that is correct

File: pipeline-runner/DESCRIPTION [hunk 2]
```diff
@@ -15,8 +15,9 @@ LazyData: true
 Roxygen: list(markdown = TRUE)
 RoxygenNote: 7.1.2
 Suggests: 
+    httptest,
     testthat (>= 3.0.0)
-Imports:
```
Comment (ivababukova) [inline pipeline-runner/DESCRIPTION:19]: what is this DESCRIPTION file for?

File: pipeline-runner/R/qc-helpers.R [hunk 3]
```diff
@@ -200,7 +200,7 @@ getClusters <- function(clustering_method, resolution, data) {
 
   if (clustering_method == "leiden") {
     # emulate FindClusters, which overwrites seurat_clusters slot and meta.data column
-    g <- getSNNiGraph(data)
+    g <- getSNNiGraph(data, active.reduction)
```
Comment (ivababukova) [inline pipeline-runner/R/qc-helpers.R:203]: Can you rename this "g" to something more meaningful? 
I see that this `getSNNiGraph` function was used before too, but it wasn't defined before. Was there a built-in function in Seurat, called `getSNNiGraph` that we were using in the past in this case?

File: pipeline-runner/R/qc-helpers.R [hunk 4]
```diff
@@ -216,6 +216,33 @@ getClusters <- function(clustering_method, resolution, data) {
...
+  # if doesn't exist, run SNN
+  if (!snn_name %in% names(data)) {
+    data <- Seurat::FindNeighbors(data, reduction = active.reduction)
+  }
+
+  # convert Seurat Graph object to igraph
+  # similar to https://github.com/joshpeters/westerlund/blob/46609a68855d64ed06f436a6e2628578248d3237/R/functions.R#L85
+  adj_matrix <-
+    Matrix::Matrix(as.matrix(data@graphs[[snn_name]]), sparse = TRUE)
+  g <- igraph::graph_from_adjacency_matrix(adj_matrix,
```
Comment (ivababukova) [inline pipeline-runner/R/qc-helpers.R:240]: in here, you can use better name for this "g".

**Review by ivababukova — APPROVED** [review]

File: pipeline-runner/DESCRIPTION (hunk 2 above)
Comment (gerbeldo) [reply pipeline-runner/DESCRIPTION:19]: It's the file that defines this directory as an R package. It contains some metadata (package name and authors) but also lists dependencies. Generally, it's not modified by hand, there are functions that take care of it when adding dependencies. [Here are some docs](https://r-pkgs.org/description.html#description) if you are interested.

File: pipeline-runner/R/qc-helpers.R (hunk 3 above)
Comment (gerbeldo) [reply pipeline-runner/R/qc-helpers.R:203]: I copied it from the worker verbatim. This function missing seems to be the reason Leiden clustering was not working in the pipeline (now the clustering tests with leiden pass). And sure, I'll rename to snn_graph.

Comment (gerbeldo) [reply pipeline-runner/R/qc-helpers.R:203]: sure! I'll rename in the worker as well

File: pipeline-runner/R/qc-helpers.R (hunk 4 above)
Comment (gerbeldo) [reply pipeline-runner/R/qc-helpers.R]: done, also renamed in worker

File: pipeline-runner/tests/testthat/test-qc-7-embed_and_cluster.R [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+  for (algo in algos) {
+    res <- stubbed_runClusters(algo, data)
+    n_cells <- nrow(res)
+
+    expect_equal(n_cells, expected_n_cells)
+  }
+})
+
+test_that("runClusters orders barcodes correctly", {
+  algos <- c("louvain", "leiden")
```
Comment (ogibson-biomage) [inline pipeline-runner/tests/testthat/test-qc-7-embed_and_cluster.R:85]:
```suggestion
test_that("runClusters returns same order of barcodes as seurat object", {
  algos <- c("louvain", "leiden")
```

**Review by ogibson-biomage — APPROVED** [review]

**Pull request merged**